ipaserver/install/ipa_migrate.py

```diff
@@ -0,0 +1,2028 @@
+                val = str(val)
+                vals.append(ensure_str(val))
+            else:
+                # Just a string
+                vals.append(val)
+
+        return self.normalize_vals(vals)
+
+    def build_ldap_entry(self, dn, attrs):
+        """
+        Take a DN and some attributes and build an LDAPEntry.  Used when
+        adding entries to the local server
+        """
+        entry = LDAPEntry(self.local_conn, DN(dn))
+        range_reset = False
+        for attr, values in attrs.items():
+            if (self.args.reset_range or self.mode == "stage-mode") and \
+                    attr.lower() in DNA_REGEN_ATTRS:
+                # Set the magic regen value
+                values = [DNA_REGEN_VAL]
+                self.log_debug(f"Resetting DNA range for new entry: {dn}")
+                range_reset = True
+            entry[attr] = values
+        if range_reset:
+            stats['reset_range'] += 1
+        return entry
+
+    def clean_entry(self, entry_dn, entry_type, entry_attrs):
+        """
+        Clean up the entry from the remote server
+
+        - Remove attributes from the ignore/strip list
+        - Reset suffix in all attributes
+        - If REALM was changed reset it to the new value
+        - Remove objectclasses from groups (IGNORE_OC list)
+        - userCertificate is removed if issued by IPA
+        - Remove unused objectclasses
+        """
+
+        # Don't clean DNS entries
+        if entry_type.startswith("dns"):
+            return entry_attrs
+
+        # Set the attrs we want to remove
+        remove_list = []
+        remove_attrs = STRIP_ATTRS + STRIP_OP_ATTRS
+        if self.args.mode != "prod-mode":
+            remove_attrs += PROD_ATTRS
+        if entry_type == "user":
+            remove_attrs.remove('ipasshpubkey')
+
+        # Walk the entry normalizing and marking attrs to remove as needed
+        for attr in entry_attrs:
+            if attr.lower() in remove_attrs:
+                remove_list.append(attr)
+                continue
+
+            # Replace suffix/realm/host/domain in all values
+            entry_attrs[attr] = self.convert_values(entry_attrs[attr])
+
+            # Check userCertificate issuer and remove IPA CA certs
+            if attr.lower() == "usercertificate" and \
+               self.remove_usercert(entry_dn, entry_attrs[attr]):
+                # This cert was issued by IPA, remove it
+                remove_list.append(attr)
+
+        # Cleanup up entry attributes
+        for remove_attr in remove_list:
+            del entry_attrs[remove_attr]
+
+        # Normalize the objectclass name -> objectClass
+        normalize_attr(entry_attrs, 'objectClass')
+
+        # Cleanup objectclasses from groups (users too?)
+        if entry_type == "group":
+            for oc in ensure_list_str(entry_attrs['objectClass']):
+                if oc.lower() in IGNORE_OC:
+                    entry_attrs['objectClass'].remove(oc.encode())
```

Comment:
For naming I wonder if IGNORE_OC should be STRIP_OC.
I can't quite get my head around how/where mepmanagedentry is removed but the resulting entries post-migration look correct.

Reply:
> For naming I wonder if IGNORE_OC should be STRIP_OC. I can't quite get my head around how/where mepmanagedentry is removed but the resulting entries post-migration look correct.

It's stripped based off comments in the original design doc :-)  I'm not sure why -  I was just following the design

Reply:
I guess its confusing that there are two lists that do the same thing. The same is true for IGNORE_ATTRS and STRIP_ATTRS.